### DISCUSSION TITLE
Is it possible to include several tables of contents within the same file?

### DISCUSSION THREAD

**Discussion by eaxca1** [typst/typst#2310] [Q&A] [answered]
Hi everyone,

I am working on a multi-file document with some chapter subfiles that are all included in a `main.typ` file. For each chapter, I would like to have a table of contents (TOC) that shows only the headings of said chapter.

When I compile the chapters separately, I get the correct TOCs by calling `outline()` (i.e., chap1.pdf, chap2.pdf, etc. all have the correct TOCs). However, when I compile `main.typ`, I always get the same TOC, one that shows all the headings of my document.

Does anybody know if there is a way to limit the scope of the `outline()` function so that it applies 'locally' to each chapter file when compiling the whole document? I tried playing a bit with the `target` parameter and `query` function, but can't get the desired result. Hopefully someone has some idea about how to approach this!

Many thanks for your impressive work.

**Comment by Dherse** [comment]
You can use the `target` argument with a query that uses the `after` and `before` functions on query: https://typst.app/docs/reference/meta/selector/#definitions-before

It would look something like: (incomplete, illustrative example)

```typc
locate(loc => {
  let next_heading = query(selector(heading).after(loc))
  outline(target: selector(heading).before(next_heading.at(0).location()).after(loc))
})
```

**Reply by Dherse** [reply, marked as answer]
My recommendation is to do the following:
```ts
#show heading.where(level: 1, outlined: true): it => [
  #it

  #set text(size: 12pt, weight: "regular")
  #outline(
      title: "Chapter outline",
      indent: true,
      target: heading
        .where(level: 1)
        .or(heading.where(level: 2))
        .after(it.location(), inclusive: true)
        .before(
          heading
            .where(level: 1, outlined: true)
            .after(it.location(), inclusive: false),
          inclusive: false,
        )
  )
]
```

**Reply by eaxca1** [reply]
That works beautifully, thanks! So `query` is not needed after all, only manipulating `target` wisely.

I think I need to do some small tweaks so it fits in my document, but it seems to do exactly what I wanted. Many thanks @Dherse.

**Reply by Dherse** [reply]
No worries, hope you have fun ;)

I kinda wish this specific example was on the doc page for outline since it's such a common use (@reknih)
